trieste/acquisition/rule.py

```diff
@@ -2256,27 +2267,36 @@ def location(self, location: TensorType) -> None:
         self._location_ix = tf.squeeze(location_ix, axis=-1)
 
     @property
-    def global_search_space(self) -> DiscreteSearchSpace:
+    def global_search_space(self) -> GeneralDiscreteSearchSpace:
         return self._global_search_space
 
     def _compute_global_distances(self) -> TensorType:
         # Helper method to compute and return pairwise distances along each axis in the
         # global search space.
+
         points = self.global_search_space.points
-        return tf.abs(
-            tf.expand_dims(points, -2) - tf.expand_dims(points, -3)
-        )  # [num_points, num_points, D]
+        if isinstance(self.global_search_space, HasOneHotEncoder):
+            # use Hamming distance for categorical spaces
+            return tf.math.reduce_sum(
+                tf.where(tf.expand_dims(points, -2) == tf.expand_dims(points, -3), 0, 1),
+                axis=-1,
+                keepdims=True,  # (keep last dim for distance calculation below)
+            )  # [num_points, num_points, 1]
+        else:
```

Comment:
I think maybe we should add more of an explanation here, as the categorical and numerical cases are slightly inconsistent.

The size of the last dimension for the numerical case is `D`, i.e. the distance in each dimension is calculated separately. Each dimesion is then separately tested against `distance` in `_get_points_within_distance` and it selects neighbors if all dimensions are within distance (i.e. `reduce_all` below).

For the categorical case the last dimension is `1` as we do a `reduce_sum`. I can see why that is, as we want to effectively do a `reduce_any` in `_get_points_within_distance`, i.e. the neighbors are selected if they are within `distance` in any dimension.

So we can add an explanation, or alternatively do the selection below and explicitly add a `reduce_any`.